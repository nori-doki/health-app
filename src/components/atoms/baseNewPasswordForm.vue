<template>
    <div class="new-password-form">
        <label for="password">Password *</label>
        <input :type="passwordInputType" id="password" name="password" required v-model="localPassword" autocomplete="new-password">
        <div class="new-password-form-password-eye password-eye">
            <i v-if="showPassword" @click="$emit('toggleShowPassword')" class="pi pi-eye"></i>
            <i v-else @click="$emit('toggleShowPassword')" class="pi pi-eye-slash"></i>
        </div>
        <ul class="new-password-form-password-requirements">
            <p>Password must contain:</p>
            <li :class="{ 'valid': passwordHasLowercase }">
                At least one lowercase 
                <span v-show="passwordHasLowercase">✅</span>
            </li>
            <li :class="{ 'valid': passwordHasUppercase }">
                At least one uppercase 
                <span v-show="passwordHasUppercase">✅</span>
            </li>
            <li :class="{ 'valid': passwordHasNumeric }">
                At least one numeric 
                <span v-show="passwordHasNumeric">✅</span>
            </li>
            <li :class="{ 'valid': passwordHasSpecialCharacter }">
                At least one special character 
                <span v-show="passwordHasSpecialCharacter">✅</span>
            </li>
            <li :class="{ 'valid': passwordIs8CharactersLong }">
                Minimum 8 characters 
                <span v-show="passwordIs8CharactersLong">✅</span>
            </li>
        </ul>
    </div>
    <div class="new-password-form">
        <label for="confirmPassword">Confirm Password *</label>
        <input :type="confirmPasswordInputType" id="confirmPassword" name="confirmPassword" required v-model="localConfirmPassword" :class="{ 'invalid': localConfirmPassword.length > 0 && !passwordsMatch }"  autocomplete="new-password">
        <div class="new-password-form-password-eye confirm-password-eye">
            <i v-if="showConfirmPassword" @click="$emit('toggleShowConfirmPassword')" class="pi pi-eye"></i>
            <i v-else @click="$emit('toggleShowConfirmPassword')" class="pi pi-eye-slash"></i>
        </div>
        <p v-show="localConfirmPassword.length > 0 && !passwordsMatch" class="new-password-form-error">Passwords do not match.</p>
    </div>
</template>

<script setup>
import { ref, watch } from 'vue';

const props = defineProps({
    password: String,
    confirmPassword: String,
    passwordInputType: String,
    confirmPasswordInputType: String,
    showPassword: Boolean,
    showConfirmPassword: Boolean,
    passwordHasLowercase: Boolean,
    passwordHasUppercase: Boolean,
    passwordHasNumeric: Boolean,
    passwordHasSpecialCharacter: Boolean,
    passwordIs8CharactersLong: Boolean,
    passwordsMatch: Boolean
});

const emit = defineEmits([
    'update:password',
    'update:confirmPassword',
    'toggleShowPassword',
    'toggleShowConfirmPassword',
]);

const localPassword = ref(props.password);
const localConfirmPassword = ref(props.confirmPassword);

watch(localPassword, (newVal) => emit('update:password', newVal));
watch(localConfirmPassword, (newVal) => emit('update:confirmPassword', newVal));
</script>

<style lang="scss">
.new-password-form {
            display: flex;
            flex-direction: column;
            margin: 0.4rem 0.8rem 0.4rem;
            min-height: 5.1rem;
            position: relative;

            label {
                margin-bottom: 0.4rem;
            }

            input {
                padding: 0.5rem;
                border: 1px solid #ccc;
                border-radius: 0.5rem;
                font-family: 'Quicksand', sans-serif;
                font-size: $font-small;
                font-weight: 500;
                min-width: 250px;

                &.invalid {
                    border-color: red;
                }
            }

            input[type="range"] {
                -webkit-appearance: none;
                appearance: none;
                background: transparent;
                cursor: pointer;
                border: none;
                color: $base-pink;
                &::-webkit-slider-runnable-track {
                    width: 100%;
                    height: 0.5rem;
                    cursor: pointer;
                    background: linear-gradient(
                        to right,
                        rgba(255, 253, 0, 1), 
                        rgba(255, 253, 0, 1), 
                        rgba(31, 251, 150, 1),
                        rgba(255, 253, 0, 1)
                        );
                    border-radius: 0.5rem;
                }
                &::-webkit-slider-thumb {
                    -webkit-appearance: none;
                    appearance: none;
                    width: 1rem;
                    height: 1rem;
                    border-radius: 50%;
                    background: $base-pink;
                    cursor: pointer;
                    margin-top: -0.25rem;
                }
            }

            &-error {
                color: red;
                font-size: $font-small;
                margin: 0.2rem;
            }

            &-password-requirements {
                font-size: $font-small;
                margin-top: 0.6rem;
                list-style-type: disc;
                padding-left: 0;

                li {
                    margin: 0.3rem 0 0.3rem 1rem;
                }

                .valid {
                    color: green;
                }
            }

            &-password-eye {
                position: absolute;
                right: 1rem;
                top: 2.2rem;
                cursor: pointer;
            }
        }
</style>