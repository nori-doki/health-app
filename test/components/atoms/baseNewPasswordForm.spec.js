import { describe, it, expect, vi } from 'vitest';
import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import baseNewPasswordForm from '@/components/atoms/baseNewPasswordForm.vue';
import { nextTick } from 'vue';

describe('baseNewPasswordForm', () => {
  const defaultProps = {
    password: '',
    confirmPassword: '',
    passwordInputType: 'password',
    confirmPasswordInputType: 'password',
    showPassword: false,
    showConfirmPassword: false,
    passwordHasLowercase: false,
    passwordHasUppercase: false,
    passwordHasNumeric: false,
    passwordHasSpecialCharacter: false,
    passwordIs8CharactersLong: false,
    passwordsMatch: true,
  }

  it('renders correctly', () => {
    const wrapper = mount(baseNewPasswordForm, { props: defaultProps })
    expect(wrapper.find('label[for="password"]').exists()).toBe(true)
    expect(wrapper.find('label[for="confirmPassword"]').exists()).toBe(true)
  })

  it('binds password and confirmPassword inputs', async () => {
    const wrapper = mount(baseNewPasswordForm, { props: defaultProps })

    const passwordInput = wrapper.find('input#password')
    await passwordInput.setValue('MyPass123!')
    expect(wrapper.emitted('update:password')).toBeTruthy()
    expect(wrapper.emitted('update:password')[0]).toEqual(['MyPass123!'])

    const confirmInput = wrapper.find('input#confirmPassword')
    await confirmInput.setValue('MyPass123!')
    expect(wrapper.emitted('update:confirmPassword')).toBeTruthy()
    expect(wrapper.emitted('update:confirmPassword')[0]).toEqual(['MyPass123!'])
  })

  it('emits toggleShowPassword and toggleShowConfirmPassword on eye icon click', async () => {
    const wrapper = mount(baseNewPasswordForm, {
      props: {
        ...defaultProps,
        showPassword: false,
        showConfirmPassword: true,
      }
    })

    let eyeIconPassword = wrapper.find('.password-eye .pi-eye-slash')
    await eyeIconPassword.trigger('click')
    expect(wrapper.emitted('toggleShowPassword')).toBeTruthy()

    await wrapper.setProps({ showPassword: true })
    await nextTick()
    eyeIconPassword = wrapper.find('.password-eye .pi-eye')
    await eyeIconPassword.trigger('click')
    expect(wrapper.emitted('toggleShowPassword')).toBeTruthy()

    let eyeIconConfirm = wrapper.find('.confirm-password-eye .pi-eye')
    await eyeIconConfirm.trigger('click')
    expect(wrapper.emitted('toggleShowConfirmPassword')).toBeTruthy()
    await wrapper.setProps({ showConfirmPassword: false })
    await nextTick()
    eyeIconConfirm = wrapper.find('.confirm-password-eye .pi-eye-slash')
    await eyeIconConfirm.trigger('click')
    expect(wrapper.emitted('toggleShowConfirmPassword')).toBeTruthy()
  })

  it('applies "valid" class based on password requirements', () => {
    const wrapper = mount(baseNewPasswordForm, {
      props: {
        ...defaultProps,
        passwordHasLowercase: true,
        passwordHasUppercase: true,
        passwordHasNumeric: true,
        passwordHasSpecialCharacter: false,
        passwordIs8CharactersLong: true,
      }
    })

    const listItems = wrapper.findAll('ul li')

    expect(listItems[0].classes()).toContain('valid')
    expect(listItems[1].classes()).toContain('valid')
    expect(listItems[2].classes()).toContain('valid')
    expect(listItems[3].classes()).not.toContain('valid')
    expect(listItems[4].classes()).toContain('valid')
  })

  it('shows error when passwords do not match', async () => {
    const wrapper = mount(baseNewPasswordForm, {
      props: {
        ...defaultProps,
        passwordsMatch: false,
      }
    })

    const confirmInput = wrapper.find('input#confirmPassword')
    await confirmInput.setValue('mismatch123')

    const errorMsg = wrapper.find('.new-password-form-error')
    expect(errorMsg.exists()).toBe(true)
    expect(errorMsg.text()).toContain('Passwords do not match.')
  })

  it('adds invalid class on confirmPassword input when mismatch', async () => {
    const wrapper = mount(baseNewPasswordForm, {
      props: {
        ...defaultProps,
        passwordsMatch: false,
      }
    })
    const passwordInput = wrapper.find('input#password')
    await passwordInput.setValue('MyPass123!')

    const confirmInput = wrapper.find('input#confirmPassword')
    await confirmInput.setValue('wrongPassword')

    expect(confirmInput.classes()).toContain('invalid')
  })
})