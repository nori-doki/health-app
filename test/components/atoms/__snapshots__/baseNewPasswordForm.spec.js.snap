// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseLogo > matches snapshot 1`] = `
"<div class="base-logo"><a>
    <p class="base-logo-black">80to<span class="base-logo-color">100</span></p>
  </a></div>"
`;
